zkml/assets/scripts/llms/gpt2_internal.py

```diff
@@ -0,0 +1,373 @@
+import torch
+import re
+import json
+import argparse
+import os
+from transformers import GPT2Model, GPT2Tokenizer
+from transformers import AutoModelForCausalLM
+from transformers.modeling_utils import Conv1D
+
+torch.set_printoptions(precision=6, sci_mode=False)
+
+# --- Argument Parsing --- #
+# This MUST be at the top level of the script to ensure `args` is globally available.
+parser = argparse.ArgumentParser(description="Run GPT-2 tiny model and dump weights and intermediate tensors.")
+parser.add_argument(
+    "--output-dir",
+    type=str,
+    default=".", # Default to current directory
+    help="Directory to save the output JSON files."
+)
+args = parser.parse_args() # This defines 'args'
+
+# Ensure output directory exists, executed early
+# This now correctly uses 'args' which should have just been defined.
+if not os.path.exists(args.output_dir):
+    os.makedirs(args.output_dir)
+    print(f"Created output directory: {args.output_dir}")
+
+# Load tokenizer and model
+tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
+# model = GPT2Model.from_pretrained("gpt2", output_hidden_states=True)
+model = AutoModelForCausalLM.from_pretrained("sshleifer/tiny-gpt2")
+
+model.eval()
+
+# Input
+input_text = "Hello"
+input_ids = tokenizer.encode(input_text, return_tensors="pt")
+
+# Run through model to get reference
+with torch.no_grad():
+    outputs = model(input_ids, output_hidden_states=True)
+    inputs_embeds = outputs.hidden_states[0] # This is wte(input_ids) + wpe(position_ids)
```

Comment:
So basically here we are doing again all the layers in the model, saving outputs and compare that the final output is the same as the one of the imported GPT2 model?